server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -936,7 +936,8 @@ public boolean isSearchable() {
     public enum Role implements Writeable, ToXContentFragment {
         DEFAULT((byte) 0, true, true),
         INDEX_ONLY((byte) 1, true, false),
-        SEARCH_ONLY((byte) 2, false, true);
+        SEARCH_ONLY((byte) 2, false, true),
+        NONE((byte) 3, false, false);
```

Comment:
I would prefer to have a new flag or method, used only for search routing. This enables that commits are still sent to the search shard and we can still do real-time GET against it. I'd like to keep the difference as small as possible, i.e., only do the searches on the indexing shard to allow us to do incremental refinements towards doing more on the search shard.

We might need two new roles. Or we can keep it inside the existing roles by having a new method `canHandleSearch` that has an argument containing the `IndexMetadata`.

This also avoids the explicit role-check in `RoutingNodes` and `PostWriteRefresh`. And ensures that `DEFAULT` is only used in regular ES (which I find appealing, though not necessarily mandatory).

Reply:
Hi @henningandersen , thanks for the quick comment! Ah, I thought of re-using the DEFAULT role, to avoid inserting a 5th new role that may become redundant in the remote future. Also, with the current approach, we keep the semantics of the two boolean flags; NONE simply means false for both. And we distinguish between DEFAULT in regular ES via the node's role.

However, I realize my approach is somewhat convoluted. If we do not mind having an extra new role, I think it'll be more explicit that way. I am not sure I prefer putting the `IndexMetadata` inside the Role, because so far it seems a Role is a "pure" enum (I mean it's 1-1 to an integer value), and if we put IndexMetadata inside it, it'll be more "heavy" (e.g., we transfer it across the wire, etc.). I think two new explicit roles are better for that reason.

I may try this approach and come back to you.

Reply:
I did not intend to put the `IndexMetadata` inside the role, it should stay an enum. But we could have a method `canHandleSearch(IndexMetadata imd)`, that can figure it out.

Reply:
Hi @henningandersen , ah I see. However, I find it a bit counter-intuitive to have a separate `canHandleSearch()` method (whether it receives an IndexMetadata argument or not). This would mean:

* `INDEX_ONLY` would then have `isSearchable()` false and `canHandleSearch()` true. This combination would tell us that it can receive search requests but not (m)gets.
* `SEARCH_ONLY` would then have `isSearchable()` false and `canHandleSearch()` false. This combination would tell us to enable (m)gets for these shards; which is even more counter-intuitive since both "search" methods of the role return false.

For this reason, for the moment I'm going with two new explicit roles that would say whether searches and (m)gets are separately allowed or not. Not sure it's definitely the better approach, but I find it less confusing. Lets review the upcoming commit and see.

Reply:
So I pushed a new commit that has the two new explicit roles.

I've thought about the alternative of a `canHandleSearch(IndexMetadata imd)` more. It would indeed minimize code changes related to new shard roles. On the other hand, we'd need to move the new stateless setting we want (so we can check whether the shard can handle search or not) to the core code base. Not sure whether we'd want that? Alternatively, we could also have some sort of plugin-based method for `canHandleSearch()` that we can override in stateless.

I am not sure if the above alternatives are more clear. For the moment, I like the current PR which is more explicit. I proposed a new flag in role called `gets` which signifies whether the shard can handle (m)gets. The two new roles we'd need are `INDEX_SEARCH` and `GETS_ONLY`.

While I was at that, I also introduced a `canHaveUnpromotables()` in the role to signify the promotable shards that may be associated with unpromotable shards, and need special handling (such as refreshing the unpromotables, etc.). I find this method better-suited for the couple of places that use it.

All of these are still debatable of course. I understand this PR/topic may have lively discussions. So shoot if you have more comments.

Reply:
To clarify more on the new flags of the roles:

* the `searchable` flag is used only in `org.elasticsearch.cluster.routing.OperationRouting#searchShards()`. This directs searches (and other operations that depend on them) to `SEARCH_ONLY` or `INDEX_SEARCH` shards.
* the new `gets` flag is used primarily in `org.elasticsearch.action.get.TransportGetAction#shards()` to direct (m)gets.

> I'd like to keep the difference as small as possible, i.e., only do the searches on the indexing shard to allow us to do incremental refinements towards doing more on the search shard.

I believe, the new `gets` flag can be used to do the incremental refinements. And finally even remove it when we do not need it anymore. I believe we may want a better name than `gets` and `canHandleGets()` to signify these search shards that we'll gradually refine. But I am not sure what name I could give to the flag. For the moment, I thought of `gets` because that's the only discriminating factor we have for these shards.